mithril-signer/src/runtime/state_machine.rs

```diff
@@ -374,12 +356,43 @@ impl StateMachine {
                 nested_error: Some(e),
             })?;
 
-        Ok(SignerState::Registered { epoch })
+        match self
+            .runner
+            .can_sign_current_epoch()
+            .await
+            .map_err(|e| RuntimeError::KeepState {
+                message: "Failed to check if signer can sign in the current epoch in 'unregistered → ?' phase".to_string(),
+                nested_error: Some(e),
+            })? {
+            true => Ok(SignerState::ReadyToSign { epoch, last_signed_entity_type: None }),
+            false => Ok(SignerState::RegisteredNotAbleToSign { epoch }),
+        }
+    }
+
+    async fn transition_from_registered_not_able_to_sign_to_unregistered(
+        &self,
+        epoch: Epoch,
+    ) -> Result<SignerState, RuntimeError> {
+        self.update_era_checker(epoch, "registered not able to sign → unregistered")
+            .await?;
+
+        Ok(SignerState::Unregistered { epoch })
     }
 
-    /// Launch the transition process from the `Registered` to the `Signed` state.
-    async fn transition_from_registered_to_signed(
+    async fn transition_from_ready_to_sign_to_unregistered(
         &self,
+        epoch: Epoch,
+    ) -> Result<SignerState, RuntimeError> {
+        self.update_era_checker(epoch, "ready to sign → unregistered")
+            .await?;
+
+        Ok(SignerState::Unregistered { epoch })
+    }
+
+    /// Launch the transition process from the `ReadyToSign` to the `ReadyToSign` state.
+    async fn transition_from_ready_to_sign_to_ready_to_sign(
```

Comment:
This transition function should only use an epoch (there is no need to use 2 different epochs) and a signed entity type (this will better prepare the removal of the pending certificate).